ICAFIX/scripts/functionhighpassandvariancenormalize.m

```diff
@@ -32,94 +34,125 @@ function functionhighpassandvariancenormalize(TR,hp,fmri,WBC,varargin)
-  pdflag=true;
-  hp=1;
-  % But in this case, preserve use of "_hp0" in the filename to indicate a linear detrend (for backwards compatibility in file naming)
-  hpstring = '_hp0';
+if length(varargin) >= 4
+    switch varargin{4}
+        case {'default', ''}
+            %leave things alone
+        case 'melodiclike'
+            iters = 1;
+            VNhalfdim = true;
+        otherwise
+            error(['unknown ICADIMMODE value: ' varargin{4}]);
+    end
 end
-if pdflag
-  if (~isscalar(hp) || hp < 0 || hp ~= fix(hp))
-	error('%s: Invalid specification for the order of the polynomial detrending', mfilename);
-  end
+
+%% Allow for compiled matlab
+if isdeployed()
+    TR = str2double(TR);
 end
 
-%% Finish setting up hpstring to use in file names
-if ~pdflag
-  pdstring = '';
+% Check whether polynomial detrend is being requested for the high-pass filtering.
+% Coincidentally deals with compiled matlab arguments
+if ischar(hp)
+    hp = lower(hp);
+    if strncmp(hp,pdstring,numel(pdstring))
+        pdflag = true;
+        polydeg = str2double(hp(numel(pdstring)+1:end));  % different purposes should use a different variable name
+        if (~isscalar(hp) || hp < 0 || hp ~= fix(hp))
+            error('%s: Invalid specification for the order of the polynomial detrending', mfilename);
+        end
+    else
+        hp = str2double(hp);
+        if ~isscalar(hp) || isnan(hp)  % Allow for hp to be provided as a string that contains purely numeric elements
+            error('%s: Invalid specification for the high-pass filter', mfilename);
+        end
+    end
 end
-if isempty(hpstring) && hp>=0
+
+%write out the hpstring logic explicitly, rather than tracing a bunch of code paths
+%must NOT be moved below the hp == 0 check below
+if pdflag
+    %explicit polynomial detrending can say whatever
     hpstring = ['_hp' pdstring num2str(hp)];
+else
+    if hp < 0
+        %negatives mean skip hp, don't put it in the filename
+        hpstring = '';
+    else
+        %hp='0' ends up here
+        hpstring = ['_hp' num2str(hp)];
+    end
 end
 
-%% Load the motion confounds, and the CIFTI (if hp>=0) (don't need either if hp<0)
-if hp>=0
+%special case hp=0 to do linear detrend
+if ~pdflag && hp == 0
+    warning('%s: hp=0 will be interpreted as polynomial detrending of order 1', mfilename);
+    pdflag = true;
+    polydeg = 1;
+end
+
+% NOTE: To interpret the logic of the following "Compute, Save, Apply" sections need to know that this function expects that: 
+% (1) individual run time series will be passed in with hp requested
+% (2) concatenated (multi-run) time series will be passed in with hp<0
+%% [This is a bit of hack: would be cleaner if we disentangled the operation of this function from the specific needs of 'hcp_fix_multi_run'].
+
+%TSC: so, let's make a new variable that actually conveys some understanding of the logic
+%HACK: assume hp < 0 is equivalent to concatenated input
```

Comment:
```suggestion
%HACK: assume hp < 0 is equivalent to concatenated input
Note that concatenated input (hp < 0) never generates a VN version of the CIFTI.
```

Reply:
Concatenated mode also doesn't save a VN file for volume (or cifti, obviously).  On the other hand, the volume VN map does get computed and used internally in concatenated mode, just not saved to a file.  So, your comment, while technically accurate, could be misinterpreted to imply that concatenated mode saves a volume VN file (which it doesn't).  Who are you trying to save from being surprised by adding this comment?

Reply:
It's for me, and anyone else trying to better (and more quickly) understand the practical difference between the two modes, without having to comb the code for 30 minutes.  I've edited my suggestion to be more precise.

Reply:
Incidentally, I've realized (consistent with your earlier point) that mods to `functionhighpass` would be necessary if wanted to use it as part of processing just a single run through FIX (ala `hcp_fix` style).  A bunch of the stuff that it is doing in `singlerun` mode is only necessary in the context of prep'ing individual runs for use in MR-FIX.  Specifically, I think all we would need/want is the filtering related stuff.  The VN stuff would be unnecessary.